### DISCUSSION TITLE
Accessing PNG compressed image data

### DISCUSSION THREAD

**Discussion by tonymarklove** [libvips/libvips#3961] [Q&A] [answered]
This is a bit of a niche use case, so I expect that the answer is "no", but is it possible to get PNG compressed image data, without it being wrapped in a PNG file format? Effectively I would like the contents of the PNG `IDAT` chunk(s) without generating an entire PNG.

The reason this is useful is that PDFs allow images to be embedded using both the PNG filtering and PNG compression algorithms, but don't want all the PNG headers. So currently I am asking libvips to generate a PNG into a buffer, parsing the chunk headers, concatenating the `IDAT` chunks together and saving the result. It would be really handy if I didn't have to make libvips perform work that I'm about to immediately undo.

_Edit_: As an aside: I believe PDFs should also support CMYK (4-channel) images with the PNG filtering and compression applied. I'm sure that would be a bigger change for libvips to support, but in a dream scenario that would be possible too. 😄

**Comment by jcupitt** [comment, marked as answer]
Hi @tonymarklove,

Sorry, I think that's a bit too specialised for libvips. It aims to be a general image processing library and doesn't have much support for the insides of specific file formats, other than trying to load and save quickly.

Have you had a look at libspng? I think it'll (probably!) do what you want:

https://github.com/randy408/libspng

Re. PDFs -- libvips uses either poppler or PDFium for PDF rendering, so PNG compression for CMYK data would need to be implemented in one of those libraries. I'd check their feature lists and see what they have.

**Reply by tonymarklove** [reply]
Hi @jcupitt,

Thanks. Yes, that's fair enough, it's a niche use case for sure.

What I didn't make clear above is that I'm actually using Ruby to process these images (using `ruby-vips`) and to generate the PDFs (using [Prawn](https://github.com/prawnpdf/prawn)), making libspng a more difficult option. But I will look into it!

Thanks for your help
